server/src/main/java/com/cloud/vm/schedule/VMScheduleManagerImpl.java

```diff
@@ -0,0 +1,598 @@
+                }
+            }
+        } finally {
+            scanLock.releaseRef();
+        }
+
+        scanLock = GlobalLock.getInternLock("vm.poll");
+        try {
+            if (scanLock.lock(5)) {
+                try {
+                    scheduleVMs();
+                } finally {
+                    scanLock.unlock();
+                }
+            }
+        } finally {
+            scanLock.releaseRef();
+        }
+    }
+
+    @DB
+    public void scheduleVMs() {
+        String displayTime = DateUtil.displayDateInTimezone(DateUtil.GMT_TIMEZONE, currentTimestamp);
+        LOGGER.debug("VM poll is being called at " + displayTime);
+
+        final List<VMScheduleVO> vmsToBeExecuted = vmScheduleDao.listAll();
+        for (final VMScheduleVO vmSchedule : vmsToBeExecuted) {
+            final long timeDifference = DateUtil.getTimeDifference(vmSchedule.getScheduledTimestamp(), currentTimestamp);
+
+            if (timeDifference <= 1) {
+                final Long vmScheduleId = vmSchedule.getId();
+                final Long vmId = vmSchedule.getVmId();
+
+                final VMInstanceVO vm = vmInstanceDao.findById(vmId);
+                if (vm == null) {
+                    vmScheduleDao.remove(vmScheduleId);
+                    continue;
+                }
+
+                if (LOGGER.isDebugEnabled()) {
+                    final Date scheduledTimestamp = vmSchedule.getScheduledTimestamp();
+                    displayTime = DateUtil.displayDateInTimezone(DateUtil.GMT_TIMEZONE, scheduledTimestamp);
+                    LOGGER.debug(String.format("Scheduling for VM [ID: %s, name: %s] schedule id: [%s] at [%s].",
+                            vm.getId(), vm.getInstanceName(), vmSchedule.getId(), displayTime));
+                }
+
+                VMScheduleVO tmpVMScheduleVO = null;
+
+                try {
+                    if (vmSchedule.getState() == VMSchedule.State.ENABLED) {
+                        tmpVMScheduleVO = vmScheduleDao.acquireInLockTable(vmScheduleId);
+                        Long jobId = performActionOnVM(vmSchedule.getAction(), vm, vmSchedule);
+                        if (jobId != null) {
+                            tmpVMScheduleVO .setAsyncJobId(jobId);
+                            vmScheduleDao.update(vmScheduleId, tmpVMScheduleVO);
+                        }
+                        scheduleNextVMJob(vmSchedule);
+                    }
+                } catch (Exception e) {
+                    LOGGER.error(String.format("Scheduling VM failed due to: [%s].", e.getMessage()), e);
+                } finally {
+                    if (tmpVMScheduleVO != null) {
+                        vmScheduleDao.releaseFromLockTable(vmScheduleId);
+                    }
+                }
+            }
+        }
+    }
+
+    private Map<String, String> setVMTag(VMScheduleVO vmSchedule) {
+        Map<String,String> Tag = new HashMap<>();
+        if (!vmSchedule.getTag().isEmpty()) {
```

Comment:
might `getTag` be null ? if so, NPE will be thrown
can use CollectionUtils instead